### PR TITLE
Subtree update of `rust-analyzer`

### PR DESCRIPTION
**Pull request by lnicola**
r? @HKalbasi

### CODE REVIEW

**Comment by hlopko** [comment]
Thank you both @lnicola and @HKalbasi for enduring :heart:

**Comment by Veykril** [comment]
https://github.com/rust-lang/rust/blob/3d0e99d632f4bbeaab979d32bd8700f170ddb6b1/src/tools/tidy/src/deps.rs#L130 needs to go since we removed anymap

**Comment by HKalbasi** [comment]
The current CI failure is https://github.com/rust-lang/rust/pull/116136#issuecomment-1733181595

**Comment by lnicola** [comment]
Now it's "error[E0463]: can't find crate for `rustc_dependencies`", but the crate is there, since it failed to compile during the previous run.

**Comment by HKalbasi** [comment]
Is it possible to run the Ci code locally? I can't reproduce this using simple `cargo check`.

**Comment by lnicola** [comment]
I tried `rm config.toml; ./x.py setup`, answered with `d` (tools) and no to the other questions, then `./x.py build src/tools/rust-analyzer/`. But I'm getting a different error:

```
error[E0463]: can't find crate for `rustc_lexer`
 --> crates/rustc-dependencies/src/lib.rs:7:1
  |
7 | extern crate rustc_lexer;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate
  |
  = help: maybe you need to install the missing components with: `rustup component add rust-src rustc-dev llvm-tools-preview`

error[E0463]: can't find crate for `rustc_parse_format`
  --> crates/rustc-dependencies/src/lib.rs:18:1
   |
18 | extern crate rustc_parse_format;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate
   |
   = help: maybe you need to install the missing components with: `rustup component add rust-src rustc-dev llvm-tools-preview`

error[E0463]: can't find crate for `rustc_abi`
  --> crates/rustc-dependencies/src/lib.rs:29:1
   |
29 | extern crate rustc_abi;
   | ^^^^^^^^^^^^^^^^^^^^^^^ can't find crate
   |
   = help: maybe you need to install the missing components with: `rustup component add rust-src rustc-dev llvm-tools-preview`

error[E0463]: can't find crate for `rustc_index`
  --> crates/rustc-dependencies/src/lib.rs:40:1
   |
40 | extern crate rustc_index;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate
   |
   = help: maybe you need to install the missing components with: `rustup component add rust-src rustc-dev llvm-tools-preview`
```

**Comment by lnicola** [comment]
Locally I see a `SYSROOT=build/x86_64-unknown-linux-gnu/stage2`, which doesn't seem to contain them:

```
libaddr2line-86ce554ff6ad08cd.rlib
libadler-bfa7a39ebb77d49a.rlib
liballoc-b30975e2b876cf32.rlib
libcfg_if-1a6fbee0083900be.rlib
libcompiler_builtins-408115ade0360d2a.rlib
libcore-ec02dd343723da85.rlib
libgetopts-1955df00aeff7d49.rlib
libgimli-1be635a4727d8e5e.rlib
libhashbrown-0d17ba1067febc2c.rlib
liblibc-f6d43cc82140783e.rlib
libLLVM-17-rust-1.75.0-nightly.so
libmemchr-795b2243f06bd098.rlib
libminiz_oxide-b3c63e8997fdca47.rlib
libobject-fe0525e327a5464a.rlib
libpanic_abort-fcb1be77671e868f.rlib
libpanic_unwind-8507c56d3a79c8f2.rlib
libproc_macro-8d2ac1ed1b679025.rlib
libprofiler_builtins-b59af8679c990a7a.rlib
librustc_demangle-1e55d89766af9c5f.rlib
librustc-nightly_rt.asan.a
librustc-nightly_rt.lsan.a
librustc-nightly_rt.msan.a
librustc-nightly_rt.safestack.a
librustc-nightly_rt.tsan.a
librustc_std_workspace_alloc-2408fbe38b247900.rlib
librustc_std_workspace_core-cd2f2bc505f56f50.rlib
librustc_std_workspace_std-ddd1fbac14da9589.rlib
libstd_detect-4314ba11477b639f.rlib
libstd-eb8a07e9874efeeb.rlib
libstd-eb8a07e9874efeeb.so
libsysroot-1bf53517571ddd84.rlib
libtest-64a23a292c81730f.rlib
libtest-64a23a292c81730f.so
libunicode_width-c46dd535c28df077.rlib
libunwind-293861b2a6629d02.rlib
```

But CI is failing at "checking stage0 rust-analyzer artifacts". So maybe:

 - a stage2 build doesn't include `rustc_lexer` and friends
 - the stage0 build seems to "check `rustc-dependencies`", but then can't find it

**Comment by HKalbasi** [comment]
> I tried rm config.toml; ./x.py setup, answered with d (tools) and no to the other questions, then ./x.py build src/tools/rust-analyzer/. But I'm getting a different error:

I was able to fix this by changing the `ToolStd` to `ToolRustc` in the `RustAnalyzer::run()` at `tool.rs`. But I'm not sure if it helps with the CI problem.

**Comment by HKalbasi** [comment]
@bjorn3 Hi, can you take a look at this?

**Comment by lnicola** [comment]
Also asked in https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/How.20to.20build.20rust-analyzer.3F.

**Comment by onur-ozkan** [comment]
> I tried rm config.toml; ./x.py setup, answered with d (tools) and no to the other questions, then ./x.py build src/tools/rust-analyzer/. But I'm getting a different error:

Here https://github.com/rust-lang/rust/actions/runs/6850648136/job/18625227820#step:26:116, you can see CI uses different build configuration.

I am able to reproduce the problem by following CI flow as much as possible.

HEAD:

```sh
  ~/devspace/onur-ozkan/rust  master $ git log -1
commit b96a0c864fdfd4a559f93f9060c1131968697ce6 (HEAD -> master, lnicola/sync-from-ra)
Author: Laurențiu Nicola <lnicola@dend.ro>
Date:   Mon Nov 13 15:27:44 2023 +0200

    Try Mode::ToolRustc
```

config.toml:

```toml
profile = "dist"

[llvm]
download-ci-llvm = true
assertions      = true

[build]

submodules = false
locked-deps = true
cargo-native-static = true

[rust]
codegen-backends = ["llvm", "cranelift"]
validate-mir-opts = 3
codegen-units-std = 1
verify-llvm-ir = true
dist-src = false
channel = "nightly"
debug-assertions = true
overflow-checks = true

[dist]
compression-profile = "balanced"
compression-formats = ['xz']
missing-tools = true
```

`x check --target=x86_64-unknown-linux-gnu --host=x86_64-unknown-linux-gnu` (replace the triple with yours if different) gives the error just like in CI.

**Comment by onur-ozkan** [comment]
fwiw when `in-rust-tree` turned off, rust-analyzer compiles without any problem.

**Comment by HKalbasi** [comment]
Progress! We now need to add `#![cfg_attr(feature = "in-rust-tree", feature(rustc_private))]` to `hir-ty` (and probably some other crates).

File: src/bootstrap/src/core/build_steps/check.rs [hunk 1]
```diff
@@ -381,7 +381,7 @@ impl Step for RustAnalyzer {
         let mut cargo = prepare_tool_cargo(
             builder,
             compiler,
-            Mode::ToolStd,
+            Mode::ToolRustc,
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/check.rs:384]: Couple lines above there is `builder.ensure(Std::new(target));`. Could you please replace it with `builder.ensure(Rustc::new(target, builder));`?

File: src/bootstrap/src/core/build_steps/test.rs [hunk 2]
```diff
@@ -378,7 +378,7 @@ impl Step for RustAnalyzer {
         let mut cargo = tool::prepare_tool_cargo(
             builder,
             compiler,
-            Mode::ToolStd,
+            Mode::ToolRustc,
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/test.rs:381]: Same here

**Review by hkalbasi — APPROVED** [review]

File: src/bootstrap/src/core/build_steps/test.rs [hunk 3]
```diff
@@ -369,7 +369,7 @@ impl Step for RustAnalyzer {
 
         // We don't need to build the whole Rust Analyzer for the proc-macro-srv test suite,
         // but we do need the standard library to be present.
-        builder.ensure(compile::Std::new(compiler, host));
+        builder.ensure(compile::Rustc::new(compiler, host));
```
Comment (lnicola) [inline src/bootstrap/src/core/build_steps/test.rs:372]: Like this? Or `target`? Or `builder`? :sweat_smile:

Comment (onur-ozkan) [reply src/bootstrap/src/core/build_steps/test.rs:372]: this seems fine

**Comment by onur-ozkan** [comment]
r=me once CI is green.

@bors rollup=never

**Comment by lnicola** [comment]
```
error: crate `rustc_lexer` required to be available in rlib format, but was not found in this form
  |
  = help: try adding `extern crate rustc_driver;` at the top level of this crate

error: crate `unicode_properties` required to be available in rlib format, but was not found in this form

error: crate `unicode_xid` required to be available in rlib format, but was not found in this form

error: crate `rustc_parse_format` required to be available in rlib format, but was not found in this form
  |
  = help: try adding `extern crate rustc_driver;` at the top level of this crate

error: crate `rustc_index` required to be available in rlib format, but was not found in this form
  |
  = help: try adding `extern crate rustc_driver;` at the top level of this crate

error: crate `arrayvec` required to be available in rlib format, but was not found in this form

error: crate `smallvec` required to be available in rlib format, but was not found in this form

error: crate `rustc_serialize` required to be available in rlib format, but was not found in this form
  |
  = help: try adding `extern crate rustc_driver;` at the top level of this crate

error: crate `thin_vec` required to be available in rlib format, but was not found in this form

error: crate `indexmap` required to be available in rlib format, but was not found in this form
```

These also show up locally with https://github.com/rust-lang/rust/commit/c2c065c3eab89a3828fc72d4fd45b70658ee08c7 applied and `cargo +nightly build --features in-rust-tree`.

**Comment by lnicola** [comment]
@HKalbasi you mean like this, right? https://github.com/rust-lang/rust/pull/117701/commits/6ad73f5ee8bc25f89966b258400056d1f776760d

Doesn't work on my machine, but let's try it.

**Comment by HKalbasi** [comment]
> Doesn't work on my machine

Running `cargo +nightly build --features in-rust-tree` in the `src/tools/rust-analyzer` works on my machine on the current branch. What error do you get?

**Comment by lnicola** [comment]
> What error do you get?

It's probably unrelated, something about a missing LLVM library.

Do I set `DIST_TRY_BUILD=0` in `src/ci/github-actions/ci.yml`? I'm inclined to r+ it directly.

**Comment by onur-ozkan** [comment]
> Do I set DIST_TRY_BUILD=0 in src/ci/github-actions/ci.yml? I'm inclined to r+ it directly.

You triggered the `try` build, but try build  doesn't check tools with `DIST_TRY_BUILD=1` which is defined in the actions file.

**Pull request merged**